tools/msu_finder.rb

```diff
@@ -0,0 +1,760 @@
+# This sceript will enumerate download links for Microsoft patches.
+#
+# Author:
+# * sinn3r
+#
+###
+
+
+msfbase = __FILE__
+while File.symlink?(msfbase)
+  msfbase = File.expand_path(File.readlink(msfbase), File.dirname(msfbase))
+end
+$:.unshift(File.expand_path(File.join(File.dirname(msfbase), '..', 'lib')))
+require 'rex'
+require 'nokogiri'
+require 'uri'
+require 'json'
+require 'optparse'
+
+module MicrosoftPatch
+
+  module SiteInfo
+    TECHNET = {
+      ip:    '157.56.148.23',
+      vhost: 'technet.microsoft.com'
+    }
+
+    MICROSOFT = {
+      ip:    '104.72.230.162',
+      vhost: 'www.microsoft.com'
+    }
+
+    GOOGLEAPIS = {
+      ip:    '74.125.28.95',
+      vhost: 'www.googleapis.com'
+    }
+  end
+
+  # This class provides whatever other classes need.
+  class Base
+
+    # Prints a debug message.
+    #
+    # @param msg [String] The message to print.
+    # @return [void]
+    def print_debug(msg='')
+      $stderr.puts "[DEBUG] #{msg}"
+    end
+
+    # Prints a status message.
+    #
+    # @param msg [String] The message to print.
+    # @return [void]
+    def print_status(msg='')
+      $stderr.puts "[*] #{msg}"
+    end
+
+    # Prints an error message.
+    #
+    # @param msg [String] The message to print.
+    # @return [void]
+    def print_error(msg='')
+      $stderr.puts "[ERROR] #{msg}"
+    end
+
+    # Prints a regular message.
+    #
+    # @param msg [String] The message to print.
+    # @return pvoid
+    def print_line(msg='')
+      $stdout.puts msg
+    end
+
+    # Sends an HTTP request with Rex.
+    #
+    # @param rhost [Hash] Information about the target host. Use MicrosoftPatch::SiteInfo.
+    # @option rhost [String] :vhost
+    # @option rhost [String] :ip IPv4 address
+    # @param opts [Hash] Information about the Rex request.
+    # @return [Rex::Proto::Http::Response]
+    def send_http_request(rhost, opts={})
+      res = nil
+
+      opts.merge!({'vhost'=>rhost[:vhost]})
+
+      print_debug("Requesting: #{opts['uri']}")
+
+      cli = Rex::Proto::Http::Client.new(rhost[:ip], 443, {}, true, 'TLS1')
+      tries = 1
+      begin
+        cli.connect
+        req = cli.request_cgi(opts)
+        res = cli.send_recv(req)
+      rescue ::EOFError, Errno::ETIMEDOUT ,Errno::ECONNRESET, Rex::ConnectionError, OpenSSL::SSL::SSLError, ::Timeout::Error => e
+        if tries < 3
+          print_error("Failed to make a request, but will try again in 5 seconds...")
+          sleep(5)
+          tries += 1
+          retry
+        else
+          raise "[x] Unable to make a request: #{e.class} #{e.message}\n#{e.backtrace * "\n"}"
+        end
+      ensure
+        cli.close
+      end
+
+      res
+    end
+  end
+
+
+  # Collects MSU download links from Technet.
+  class PatchLinkCollector < Base
+
+    # Returns a response of an advisory page.
+    #
+    # @param msb [String] MSB number in this format: msxx-xxx
+    # @return [Rex::Proto::Http::Response]
+    def download_advisory(msb)
```

Comment:
I'd say send_http_request can also raise exceptions, which isn't documented. Same applies for other methods just calling send_http_request